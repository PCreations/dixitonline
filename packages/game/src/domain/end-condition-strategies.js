const remainingTurnsEndingCondition = remainingTurns => ({
  remainingTurns,
  isGameEnded: remainingTurns === -1,
});

export const defaultEndingConditionStrategy = game => {
  const playersCount = game.players.length + 1;
  const cardsCount = game.cards.length;
  const cardsPerPlayers = cardsCount / playersCount;
  return remainingTurnsEndingCondition(cardsPerPlayers <= 0 ? -1 : Math.floor(cardsCount / playersCount));
};

export const xTimesStorytellerEndingConditionsStrategy = game => {
  const playersCount = game.players.length + 1;
  const xTimesLimit = game.endCondition.xTimesStorytellerLimit;
  const currentTurnNumber = game.currentTurn.number;
  return remainingTurnsEndingCondition(playersCount * xTimesLimit - currentTurnNumber);
};

export const scoreLimitEndingCondition = game => ({
  isGameEnded: Object.values(game.score).some(s => s >= game.endCondition.scoreLimit),
});

export const getEndingConditionStrategy = game => {
  if (game.endCondition.scoreLimit) {
    return scoreLimitEndingCondition;
  }
  if (game.endCondition.xTimesStorytellerLimit) {
    return xTimesStorytellerEndingConditionsStrategy;
  }
  return defaultEndingConditionStrategy;
};
