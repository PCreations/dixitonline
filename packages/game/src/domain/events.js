export const types = {
  NEW_GAME_CREATED: '[lobby] - a new game has been created',
  PLAYER_JOINED: '[lobby] - a new player has joined a game',
  GAME_STARTED: '[lobby] - a new game has started',
  HANDS_COMPLETED: '[game] - the hands has been completed',
  GAME_ENDED: '[game] - a game has ended',
};

export const newGameCreatedEvent = ({ gameId }) => ({
  type: types.NEW_GAME_CREATED,
  payload: {
    gameId,
  },
});

export const playerJoinedGame = ({ gameId, playerId }) => ({
  type: types.PLAYER_JOINED,
  payload: {
    gameId,
    playerId,
  },
});

export const newGameStartedEvent = ({ gameId, playerIds, useAllDeck }) => ({
  type: types.GAME_STARTED,
  payload: {
    gameId,
    playerIds,
    useAllDeck,
  },
});

export const handsCompletedEvent = ({ gameId, handsByPlayerId, previousTurnId }) => ({
  type: types.HANDS_COMPLETED,
  payload: {
    gameId,
    handsByPlayerId,
    previousTurnId,
  },
});

export const gameEndedEvent = ({ gameId }) => ({
  type: types.GAME_ENDED,
  payload: {
    gameId,
  },
});
