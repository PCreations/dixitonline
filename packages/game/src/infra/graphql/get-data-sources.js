import { makeGameRepository } from '../../repos/game-repository';

export const makeGetDataSources = ({ gameRepository = makeGameRepository() } = {}) => () => ({
  gameRepository,
});
