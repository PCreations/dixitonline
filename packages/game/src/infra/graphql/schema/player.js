import { objectType } from 'nexus';

export const Player = objectType({
  name: 'GamePlayer',
  definition(t) {
    t.id('id');
    t.string('name');
    t.int('score', {
      resolve(player) {
        return player.score ?? 0;
      },
    });
  },
});
