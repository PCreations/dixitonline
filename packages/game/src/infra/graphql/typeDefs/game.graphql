type Game {
  id: ID!
  host: GamePlayer!
  players: [GamePlayer!]!
}
