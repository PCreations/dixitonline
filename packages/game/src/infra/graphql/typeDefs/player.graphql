type GamePlayer {
  id: ID!
  name: String!
}
