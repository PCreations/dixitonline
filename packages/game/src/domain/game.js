/* eslint-disable max-classes-per-file */
import { shuffle as shuffleWithSeed } from 'shuffle-seed';
import {
  newGameCreatedEvent,
  playerJoinedGame,
  newGameStartedEvent,
  handsCompletedEvent,
  gameEndedEvent,
} from './events';
import { getEndingConditionStrategy } from './end-condition-strategies';
import { equals as playerEquals } from './player';
import { makeGameResult as baseMakeGameResult, makeErrorResult } from './game-result';

const defaultShuffle = toShuffle => shuffleWithSeed(toShuffle, 'default-seed');

export const DEFAULT_END_CONDITION = {
  isGameEnded: false,
};

export const CARDS_NOT_DEALT_YET = null;

export const GameStatus = {
  WAITING_FOR_PLAYERS: 'WAITING_FOR_PLAYERS',
  STARTED: 'STARTED',
  ENDED: 'ENDED',
};

export const GameError = {
  GAME_ALREADY_JOINED: 'GAME_ALREADY_JOINED',
  GAME_ALREADY_STARTED: 'GAME_ALREADY_STARTED',
  MAXIMUM_NUMBER_OF_PLAYERS_REACHED: 'MAXIMUM_NUMBER_OF_PLAYERS_REACHED',
  ONLY_HOST_CAN_START_GAME: 'ONLY_HOST_CAN_START_GAME',
  NOT_ENOUGH_PLAYERS: 'NOT_ENOUGH_PLAYERS',
  X_TIMES_STORYTELLER_CANT_BE_LESS_THAN_ONE: 'X_TIMES_STORYTELLER_CANT_BE_LESS_THAN_ONE',
  SCORE_LIMIT_CANT_BE_LESS_THAN_ONE: 'SCORE_LIMIT_CANT_BE_LESS_THAN_ONE',
};

export const MAXIMUM_NUMBER_OF_PLAYERS = 6;
export const MINIMUM_NUMBER_OF_PLAYERS = 3;
export const NUMBER_OF_CARDS_IN_A_DECK = 84;
export const NUMBER_OF_CARDS_BY_HAND = 6;

export const makeNullCards = () => ({
  isNullCards: true,
  length: 0,
});

export const makeGame = ({
  id,
  host,
  cards = makeNullCards(),
  drawPile = [],
  score = {},
  players = [],
  status = GameStatus.WAITING_FOR_PLAYERS,
  currentTurn = { id: null, storytellerId: null, number: 0 },
  endCondition = DEFAULT_END_CONDITION,
} = {}) => {
  if (!id) throw new Error('Game must contain an id');
  if (!host) throw new Error('Game must have an host');
  if (
    typeof endCondition.xTimesStorytellerLimit === 'undefined' &&
    typeof endCondition.scoreLimit === 'undefined' &&
    typeof endCondition.isGameEnded === 'undefined'
  )
    throw new Error(`Invalid end condition, received "${endCondition}"`);

  console.log(`creating a game with ${cards.length} cards`);

  return Object.freeze({
    id,
    host,
    players,
    cards,
    drawPile,
    endCondition,
    score,
    status,
    currentTurn,
  });
};

const haveCardsNotBeenDealtYet = game => game.cards.isNullCards;

const applyEndingConditionStrategy = ({ value: game, events }) => {
  if (haveCardsNotBeenDealtYet(game)) return { value: game, events };
  const endingCondition = getEndingConditionStrategy(game);
  const { isGameEnded } = endingCondition(game);
  if (isGameEnded)
    return {
      value: makeGame({ ...game, status: GameStatus.ENDED }),
      events: events.concat(gameEndedEvent({ gameId: game.id })),
    };
  return { value: game, events };
};

const makeGameResult = (game, events = []) => applyEndingConditionStrategy(baseMakeGameResult(game, events));

export const getNumberOfCardsByHand = game =>
  getAllPlayers(game).length === 3 ? NUMBER_OF_CARDS_BY_HAND + 1 : NUMBER_OF_CARDS_BY_HAND;

export const getAllPlayers = game => [game.host, ...game.players];

export const getEndCondition = game => {
  const allPlayersLength = getAllPlayers(game).length;
  if (typeof game.endCondition.xTimesStorytellerLimit !== 'undefined') {
    return {
      remainingTurns: allPlayersLength * game.endCondition.xTimesStorytellerLimit - game.currentTurn.number,
    };
  }
  if (typeof game.endCondition.scoreLimit !== 'undefined') {
    return {
      scoreLimit: game.endCondition.scoreLimit,
    };
  }
  return {
    remainingTurns: Math.floor(game.cards.length / allPlayersLength),
  };
};

const isGameFull = game => getAllPlayers(game).length === MAXIMUM_NUMBER_OF_PLAYERS;

export const createGame = ({ gameId, host, endCondition }) => {
  if (endCondition?.xTimesStorytellerLimit !== undefined && endCondition.xTimesStorytellerLimit < 1) {
    return makeErrorResult(GameError.X_TIMES_STORYTELLER_CANT_BE_LESS_THAN_ONE);
  }
  if (endCondition?.scoreLimit !== undefined && endCondition.scoreLimit < 1) {
    return makeErrorResult(GameError.SCORE_LIMIT_CANT_BE_LESS_THAN_ONE);
  }
  return makeGameResult(makeGame({ id: gameId, host, endCondition }), [newGameCreatedEvent({ gameId })]);
};

export const joinPlayer = (game, player) => {
  if (getAllPlayers(game).some(playerEquals.bind(null, player))) return makeErrorResult(GameError.GAME_ALREADY_JOINED);
  if (isGameFull(game)) return makeErrorResult(GameError.MAXIMUM_NUMBER_OF_PLAYERS_REACHED);
  return makeGameResult(makeGame({ ...game, players: game.players.concat(player) }), [
    playerJoinedGame({ gameId: game.id, playerId: player.id }),
  ]);
};

export const startGame = (game, player) => {
  if (game.status === GameStatus.STARTED) {
    return makeErrorResult(GameError.GAME_ALREADY_STARTED);
  }
  if (getAllPlayers(game).length < MINIMUM_NUMBER_OF_PLAYERS) {
    return makeErrorResult(GameError.NOT_ENOUGH_PLAYERS);
  }
  if (playerEquals(game.host, player)) {
    const isDefaultEndCondition =
      game.endCondition?.xTimesStorytellerLimit === undefined && game.endCondition?.scoreLimit === undefined;
    return makeGameResult(
      makeGame({
        ...game,
        status: GameStatus.STARTED,
      }),
      [
        newGameStartedEvent({
          gameId: game.id,
          playerIds: getAllPlayers(game).map(({ id }) => id),
          useAllDeck: !isDefaultEndCondition,
        }),
      ]
    );
  }
  return makeErrorResult(GameError.ONLY_HOST_CAN_START_GAME);
};

export const updateDeck = (game, { discardedCards, shuffle = defaultShuffle } = {}) => {
  if (game.cards.length < getAllPlayers(game).length && game.endCondition !== DEFAULT_END_CONDITION) {
    const drawPile = shuffle([...game.drawPile, ...discardedCards, ...game.cards]);
    return makeGameResult(
      makeGame({
        ...game,
        cards: drawPile,
        drawPile: [],
      })
    );
  }
  return makeGameResult(
    makeGame({
      ...game,
      drawPile: game.drawPile.concat(discardedCards),
    })
  );
};

export const completeHands = (game, { cards, actualHandsByPlayerId, previousTurnId }) => {
  const allPlayers = getAllPlayers(game);
  const numberOfCardsByHand = getNumberOfCardsByHand(game);
  const actualCards = cards ?? game.cards;
  if (!actualHandsByPlayerId) {
    const handsByPlayerId = allPlayers.reduce(
      (hands, player, playerIndex) => ({
        ...hands,
        [player.id]: actualCards.slice(
          playerIndex * numberOfCardsByHand,
          playerIndex * numberOfCardsByHand + numberOfCardsByHand
        ),
      }),
      {}
    );
    const remainingCards = actualCards.slice(allPlayers.length * numberOfCardsByHand);
    return makeGameResult(
      makeGame({
        ...game,
        cards: remainingCards,
      }),
      [handsCompletedEvent({ gameId: game.id, handsByPlayerId })]
    );
  }
  const handsByPlayerId = Object.entries(actualHandsByPlayerId).reduce(
    (completedHands, [playerId, hand], playerIndex) => {
      const numberOfCardsToCompleteHand = numberOfCardsByHand - hand.length;
      return {
        ...completedHands,
        [playerId]: hand.concat(actualCards.slice(playerIndex, playerIndex + numberOfCardsToCompleteHand)),
      };
    },
    {}
  );
  return makeGameResult(
    makeGame({
      ...game,
      cards: actualCards.slice(allPlayers.length),
    }),
    [handsCompletedEvent({ gameId: game.id, handsByPlayerId, previousTurnId })]
  );
};

export const updateScore = (game, turnScore) => {
  const newScore = Object.entries(turnScore).reduce(
    (scores, [playerId, playerTurnScore]) => ({
      ...scores,
      [playerId]: playerTurnScore + (game.score[playerId] || 0),
    }),
    {}
  );
  console.log('updated score', newScore);
  return makeGameResult(
    makeGame({
      ...game,
      score: newScore,
    })
  );
};

export const setCurrentTurn = (game, currentTurn) =>
  makeGameResult(
    makeGame({
      ...game,
      currentTurn: {
        ...currentTurn,
        number: game.currentTurn.number + 1,
      },
    })
  );
