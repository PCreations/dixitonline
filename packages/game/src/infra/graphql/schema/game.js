import { objectType, enumType, unionType } from 'nexus';
import { Player } from './player';
import { getAllPlayers, getEndCondition, GameStatus as DomainGameStatus } from '../../../domain/game';

export const GameStatus = enumType({
  name: 'GameStatus',
  members: DomainGameStatus,
});

export const RemainingTurnsEndCondition = objectType({
  name: 'GameRemainingTurnsEndCondition',
  definition(t) {
    t.int('remainingTurns');
  },
});

export const ScoreLimitEndCondition = objectType({
  name: 'GameScoreLimitEndCondition',
  definition(t) {
    t.int('scoreLimit');
  },
});

export const GameEndCondition = unionType({
  name: 'GameEndCondition',
  definition(t) {
    t.members(RemainingTurnsEndCondition, ScoreLimitEndCondition);
    t.resolveType(obj =>
      typeof obj.remainingTurns === 'undefined' ? 'GameScoreLimitEndCondition' : 'GameRemainingTurnsEndCondition'
    );
  },
});

export const Game = objectType({
  name: 'Game',
  definition(t) {
    t.id('id');
    t.field('host', { type: Player });
    t.field('status', { type: GameStatus });
    t.id('currentTurnId', {
      nullable: true,
      resolve({ currentTurn }) {
        return currentTurn.id;
      },
    });
    t.int('remainingTurns', {
      deprecation: 'use endCondition instead',
    });
    t.field('endCondition', {
      type: GameEndCondition,
      resolve(game) {
        console.log(getEndCondition(game));
        return getEndCondition(game);
      },
    });
    t.list.field('players', {
      type: Player,
      resolve(game) {
        const { score } = game;
        const players = getAllPlayers(game).map(player => {
          return {
            ...player,
            score: score[player.id],
          };
        });
        return players;
      },
    });
  },
});
