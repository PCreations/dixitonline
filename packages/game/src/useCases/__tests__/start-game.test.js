import gql from 'graphql-tag';
import { createTestClient } from 'apollo-server-testing';
import { makeTestServer } from '../../__tests__/test-server';
import { makeGetDataSources } from '../../infra/graphql/get-data-sources';
import { buildTestGame } from '../../__tests__/dataBuilders/game';
import { buildgameRepositoryInitialGames } from '../../__tests__/dataBuilders/game-repository-initial-games';
import { makeNullGameRepository } from '../../repos';
import { buildTestPlayer } from '../../__tests__/dataBuilders/player';
import { newGameStartedEvent } from '../../domain/events';
import { getAllPlayers, GameStatus } from '../../domain/game';

const getStartGameTestServer = ({ currentUserId, currentUserUsername, numberOfPlayers = 3 } = {}) => {
  const game = buildTestGame()
    .withId('g42')
    .withXPlayers(numberOfPlayers)
    .build();
  const initialGames = buildgameRepositoryInitialGames()
    .withGames([game])
    .build();
  const gameRepository = makeNullGameRepository({
    gamesData: initialGames,
  });
  const dispatchDomainEvents = jest.fn();
  const server = makeTestServer({
    getDataSources: makeGetDataSources({
      gameRepository,
    }),
    dispatchDomainEvents,
    currentUserId: currentUserId || game.host.id,
    currentUserUsername: currentUserUsername || game.host.name,
  });

  return {
    game,
    gameRepository,
    dispatchDomainEvents,
    server,
  };
};

describe('start game', () => {
  test('a game with more than 2 players can be started by the host', async () => {
    // arrange
    const { game, gameRepository, dispatchDomainEvents, server } = getStartGameTestServer();
    const GAME_START_GAME = gql`
      mutation GameStartGame($startGameInput: GameStartGameInput!) {
        gameStartGame(startGameInput: $startGameInput) {
          ... on GameStartGameResultSuccess {
            game {
              id
            }
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: GAME_START_GAME,
      variables: { startGameInput: { gameId: 'g42' } },
      operationName: 'GameStartGame',
    });

    // assert
    expect.assertions(3);
    expect(response).toMatchSnapshot();
    const editedGame = await gameRepository.getGameById('g42');
    expect(editedGame.status).toEqual(GameStatus.STARTED);
    expect(dispatchDomainEvents).toHaveBeenCalledWith([
      newGameStartedEvent({ gameId: 'g42', playerIds: getAllPlayers(game).map(({ id }) => id), useAllDeck: false }),
    ]);
  });
  test("a player can't start a game if she's not the host", async () => {
    // arrange
    const randomPlayer = buildTestPlayer().build();
    const { game, gameRepository, dispatchDomainEvents, server } = getStartGameTestServer({
      currentUserId: randomPlayer.id,
      currentUserUsername: randomPlayer.name,
    });
    const GAME_START_GAME = gql`
      mutation GameStartGame($startGameInput: GameStartGameInput!) {
        gameStartGame(startGameInput: $startGameInput) {
          ... on GameStartGameResultError {
            type
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: GAME_START_GAME,
      variables: { startGameInput: { gameId: 'g42' } },
      operationName: 'GameStartGame',
    });

    // assert
    expect.assertions(3);
    expect(response).toMatchSnapshot();
    expect(dispatchDomainEvents).not.toHaveBeenCalled();
    await expect(gameRepository.getGameById('g42')).resolves.toEqual(game);
  });
  test("a player can't start a game if there is not enough player", async () => {
    // arrange
    const { game, gameRepository, dispatchDomainEvents, server } = getStartGameTestServer({
      numberOfPlayers: 1,
    });
    const GAME_START_GAME = gql`
      mutation GameStartGame($startGameInput: GameStartGameInput!) {
        gameStartGame(startGameInput: $startGameInput) {
          ... on GameStartGameResultError {
            type
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: GAME_START_GAME,
      variables: {
        startGameInput: {
          gameId: 'g42',
        },
      },
      operationName: 'GameStartGame',
    });

    // assert
    expect.assertions(3);
    expect(response).toMatchSnapshot();
    expect(dispatchDomainEvents).not.toHaveBeenCalled();
    await expect(gameRepository.getGameById('g42')).resolves.toEqual(game);
  });
});
