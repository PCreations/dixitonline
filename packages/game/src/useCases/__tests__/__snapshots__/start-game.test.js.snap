// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start game a game with more than 2 players can be started by the host 1`] = `
Object {
  "data": Object {
    "gameStartGame": Object {
      "game": Object {
        "id": "g42",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`start game a player can't start a game if she's not the host 1`] = `
Object {
  "data": Object {
    "gameStartGame": Object {
      "type": "ONLY_HOST_CAN_START_GAME",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`start game a player can't start a game if there is not enough player 1`] = `
Object {
  "data": Object {
    "gameStartGame": Object {
      "type": "NOT_ENOUGH_PLAYERS",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
